driver-core/src/main/java/com/datastax/driver/core/Connection.java

```diff
@@ -405,9 +405,52 @@ private void incrementAuthErrorMetric() {
         }
     }
 
-    private UnsupportedProtocolVersionException unsupportedProtocolVersionException(ProtocolVersion triedVersion, ProtocolVersion serverProtocolVersion) {
-        logger.debug("Got unsupported protocol version error from {} for version {} server supports version {}", address, triedVersion, serverProtocolVersion);
-        return new UnsupportedProtocolVersionException(address, triedVersion, serverProtocolVersion);
+    private static final Pattern SUPPORTED_VERSIONS_PATTERN = Pattern.compile("supported versions are \\(([^)]+?)\\)");
```

Comment:
> I guess what you mean is if the server version differs from the client one, then use that version, otherwise downgrade to N-1, is that right? The only problem I see with the "-1" approach is that downgrading when the server replies with the same version is not useful at all (any lower version will fail as well – what we need in that case is actually upgrade).

Unless i'm misunderstanding, we would never upgrade the protocol version, only downgrade.

I think the most straightforward approach is what @jorgebay came up with in datastax/nodejs-driver#204.  That implementation is similar to what you said with a few caveats:

    1. if the server version differs from the client one, then use that version;
    2. otherwise downgrade from the previous protocol version you tried

Steps 1 and 2 would then be repeated until a suitable protocol version is found.

I think all the other drivers (other than C# and nodejs) only do step 2.

'downgrade' could mean different things depending on the context of the protocol version, so would be nice to have a method on `ProtocolVersion` for asking for the lower protocol version.